exercise-book/src/iterators.md

```diff
@@ -0,0 +1,359 @@
+
+If you need it, we have provided a [complete solution](../../exercise-solutions/iterators/src/bin/iterators1.rs) for this exercise.
+
+## Knowledge
+
+### Iterators and iterator chains
+
+Iterators are a way to chain function calls instead of writing elaborate for loops.
+
+This lets us have a type safe way of composing control flow together by calling the right functions.
+
+For example, to double every number given by a vector, you could write a for loop:
+
+```rust
+let v = [10, 20, 30];
+let mut xs = [0, 0, 0];
+
+for idx in 0..=v.len() {
+  xs[idx] = 2 * v[idx];
+}
+```
+
+In this case, the idea of the procedure `2 * v[idx]` and indexing over the entire collection is called a [map](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.map). An idiomatic Rustacean would write something similar to the following (period indented) code:
+
+```rust
+let v = [10, 20, 30];
+let xs: Vec<_> = v.iter()
+                  .map(|elem| elem * 2)
+                  .collect();
+```
+
+No win for brevity, but it has several benefits:
+
+- Changing the underlying logic is more robust
+- Less indexing operations means you will fight the borrow checker less in the long run
+- You can parallelize your code with minimal changes using [rayon](https://crates.io/crates/rayon).
+
+The first point is not in vain - the original snippet has a bug in the upper bound, since `0..=v.len()` is inclusive!
+
+Think of iterators as lazy functions - they only carry out computation when a *consuming adapter* like `.collect()` is called, not the `.map()` itself.
+
+### Iterator chains workflow advice
+
+Start every iterator call on a new line, so that you can see closure arguments and type hints for the iterator at the end of the line clearly.
+
+When in doubt, write `.map(|x| x)` first to see what item types you get and decide on what iterator methods to use and what to do inside a closure based on that.
+
+### Turbo fish syntax `::<>`
+
+Iterators sometimes struggle to figure out the types of all intermediate steps and need assistance.
+
+```rust
+let numbers: Vec<_> = ["1", "2", "3"]
+    .iter()
+    .map(|s| s.parse::<i32>().unwrap())
+    // a turbofish in the `parse` call above
+    // helps a compiler determine the type of `n` below
+    .map(|n| n + 1)
+    .collect();
+```
+
+This `::<SomeType>` syntax is called the [turbo fish operator](https://doc.rust-lang.org/book/appendix-02-operators.html?highlight=turbo%20fish#non-operator-symbols), and it disambiguates calling the same method with different output types, like `.parse::<i32>()` and `.parse::<f64>()` (try it!)
+
+### Dealing with `.unwrap()`s in iterator chains
+
+When starting out with iterators, it's very easy to be "led astray" by locally useful `.unwrap()`s as suggested by the compiler.
+
+It's easy to get a slogging first solution with a lot of `Option` and `Result` wrapping and unwrapping that other languages wouldn't make explicit.
+
+Concretely, the following snippet:
+
+```rust
```

Comment:
This snippet and the "solution" imo go too far and essentially solve the overall exercise. How about we only show one `Result`-producing closure?

```rust
    let numbers: Vec<_> = ["1", "2", "3"]
        .iter()
        .map(|s| s.parse::<i32>())
        .filter(|r| r.is_ok())
        .map(|r| r.expect("all `Result`s are Ok here"))
        .collect();

    let numbers: Vec<_> = ["1", "2", "3"]
        .iter()
        .filter_map(|s| s.parse::<i32>().ok())
        .collect();
```

Reply:
Correct, that seems better.